source/slang/slang-intrinsic-expand.cpp

```diff
@@ -462,33 +484,41 @@ const char* IntrinsicExpandContext::_emitSpecial(const char* cursor)
             SLANG_RELEASE_ASSERT(m_argCount >= 1);
 
             auto textureArg = m_args[0].get();
+            IRType* elementType  = nullptr;
+
             if (auto baseTextureType = as<IRTextureType>(textureArg->getDataType()))
             {
-                auto elementType = baseTextureType->getElementType();
-                if (auto basicType = as<IRBasicType>(elementType))
-                {
-                    // A scalar result is expected
-                    m_writer->emit(".x");
-                }
-                else if (auto vectorType = as<IRVectorType>(elementType))
-                {
-                    // A vector result is expected
-                    auto elementCount = getIntVal(vectorType->getElementCount());
+                elementType = baseTextureType->getElementType();
+            }
+            else if(auto baseTextureSamplerType = as<IRTextureSamplerType>(textureArg->getDataType()))
+            {
+                elementType = baseTextureSamplerType->getElementType();
+            }
+            else
+            {
+                SLANG_UNEXPECTED("bad format in intrinsic definition");
+            }
 
-                    if (elementCount < 4)
-                    {
-                        char const* swiz[] = { "", ".x", ".xy", ".xyz", "" };
-                        m_writer->emit(swiz[elementCount]);
-                    }
-                }
-                else
+            SLANG_ASSERT(elementType);
+            if (auto basicType = as<IRBasicType>(elementType))
+            {
+                // A scalar result is expected
+                m_writer->emit(".x");
+            }
+            else if (auto vectorType = as<IRVectorType>(elementType))
+            {
+                // A vector result is expected
+                auto elementCount = getIntVal(vectorType->getElementCount());
+
+                if (elementCount < 4)
                 {
-                    // What other cases are possible?
+                    char const* swiz[] = { "", ".x", ".xy", ".xyz", "" };
+                    m_writer->emit(swiz[elementCount]);
                 }
             }
             else
             {
-                SLANG_UNEXPECTED("bad format in intrinsic definition");
```

Comment:
Having the 'assert' doesn't seem unreasonable.

Reply:
Yeah, I think this was an example of me trying to do defensive coding. There is in principle the possibility somebody does:

```
struct MyStuff { float4 f; }
Texture2D<MyStuff> gMyStuffTex;
```

But realistically I don't expect the rest of the Slang compiler to handle that robustly anyway. It is something that I think historically fxc allowed, but that was partly due to the "anything goes" rules fxc seemed to employ.

It is probably healthier for us to say that texel types are always scalar basic types or vectors (which is what that assert enforces, in a roundabout way), and then if we want to relax that later actually put in the necessary design effort.